src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpIsAndCastCheckTests.cs

```diff
@@ -0,0 +1,456 @@
+                new CSharpIsAndCastCheckCodeFixProvider());
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task InlineTypeCheck1()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            [|var|] v = (string)x;
+        } 
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        if (x is string v)
+        {
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingInCSharp6()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            [|var|] v = (string)x;
+        } 
+    }
+}", parseOptions: CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.CSharp6));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingInWrongName()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            [|var|] v = (string)y;
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingInWrongType()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            [|var|] v = (bool)x;
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingOnMultiVar()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            var [|v|] = (string)x, v1 = "";
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingOnNonDeclaration()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x is string)
+        {
+            [|v|] = (string)x;
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestMissingOnAsExpression()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        if (x as string)
+        {
+            [|var|] v = (string)x;
+        } 
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task InlineTypeCheckComplexExpression1()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        if ((x ? y : z) is string)
+        {
+            [|var|] v = (string)(x ? y : z);
+        } 
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        if ((x ? y : z) is string v)
```

Comment:
Hmmm... interesting example. I wonder if we should be using the helper to always parenthesize the expression that gets moved into the if-condition and mark it with a simplification annotation.